File: sdks/python/kyuubiki_sdk/operator_tasks.py
from __future__ import annotations

from typing import Any

_RECEIPT_KEYS = ("failure_receipt", "operator_task_failure_receipt")
_RECEIPT_SCHEMAS = (
    "kyuubiki.headless-operator-task-failure/v1",
    "kyuubiki.agent-operator-task-failure/v1",
    "kyuubiki.control-plane-operator-task-failure/v1",
)


def extract_operator_task_failure_receipts(payload: Any) -> list[dict[str, Any]]:
    """Return all Operator TaskIR failure receipts nested in a response payload."""

    receipts: list[dict[str, Any]] = []
    _collect_failure_receipts(payload, receipts)
    return _unique_receipts(receipts)


def operator_task_failure_actions(payload: Any) -> list[str]:
    """Return unique recovery actions from receipts and resume plans."""

    actions: list[str] = []
    for receipt in extract_operator_task_failure_receipts(payload):
        _append_recovery_action(actions, receipt.get("recovery", {}).get("required_action"))
    _collect_recovery_actions(payload, actions)
    return actions


def operator_task_recovery_summary(payload: Any) -> dict[str, Any]:
    """Return a compact recovery summary for checkpoint/resume automation."""

    receipts = extract_operator_task_failure_receipts(payload)
    return {
        "next_action": _first_string_value(payload, "next_action"),
        "target_case_ids": _first_string_list_value(payload, "target_case_ids"),
        "blocked_case_ids": _first_string_list_value(payload, "blocked_case_ids"),
        "recovery_actions": operator_task_failure_actions(payload),
        "failure_receipt_count": len(receipts),
        "failure_receipts": receipts,
    }


def _collect_failure_receipts(value: Any, receipts: list[dict[str, Any]]) -> None:
    if isinstance(value, dict):
        for key in _RECEIPT_KEYS:
            candidate = value.get(key)
            if _is_failure_receipt(candidate):
                receipts.append(candidate)
        if _is_failure_receipt(value):
            receipts.append(value)
        for child in value.values():
            _collect_failure_receipts(child, receipts)
    elif isinstance(value, list):
        for child in value:
            _collect_failure_receipts(child, receipts)


def _collect_recovery_actions(value: Any, actions: list[str]) -> None:
    if isinstance(value, dict):
        candidates = value.get("recovery_actions")
        if isinstance(candidates, list):
            for action in candidates:
                _append_recovery_action(actions, action)
        for child in value.values():
            _collect_recovery_actions(child, actions)
    elif isinstance(value, list):
        for child in value:
            _collect_recovery_actions(child, actions)


def _append_recovery_action(actions: list[str], action: Any) -> None:
    if isinstance(action, str) and action and action not in actions:
        actions.append(action)


def _first_string_value(value: Any, key: str) -> str | None:
    if isinstance(value, dict):
        candidate = value.get(key)
        if isinstance(candidate, str) and candidate:
            return candidate
        for child in value.values():
            found = _first_string_value(child, key)
            if found is not None:
                return found
    elif isinstance(value, list):
        for child in value:
            found = _first_string_value(child, key)
            if found is not None:
                return found
    return None


def _first_string_list_value(value: Any, key: str) -> list[str]:
    if isinstance(value, dict):
        candidate = value.get(key)
        if isinstance(candidate, list):
            return [item for item in candidate if isinstance(item, str)]
        for child in value.values():
            found = _first_string_list_value(child, key)
            if found:
                return found
    elif isinstance(value, list):
        for child in value:
            found = _first_string_list_value(child, key)
            if found:
                return found
    return []


def _is_failure_receipt(value: Any) -> bool:
    return isinstance(value, dict) and value.get("schema_version") in _RECEIPT_SCHEMAS


def _unique_receipts(receipts: list[dict[str, Any]]) -> list[dict[str, Any]]:
    unique: list[dict[str, Any]] = []
    seen: set[tuple[Any, ...]] = set()
    for receipt in receipts:
        key = (
            receipt.get("schema_version"),
            receipt.get("failure_stage"),
            receipt.get("reason_code"),
            receipt.get("task_id"),
            receipt.get("operator_id"),
            receipt.get("task_digest"),
        )
        if key in seen:
            continue
        seen.add(key)
        unique.append(receipt)
    return unique

File: sdks/elixir/lib/kyuubiki_sdk/operator_tasks.ex
defmodule KyuubikiSdk.OperatorTasks do
  @moduledoc "Helpers for Operator TaskIR response envelopes."

  @receipt_keys ["failure_receipt", "operator_task_failure_receipt"]
  @receipt_schemas [
    "kyuubiki.headless-operator-task-failure/v1",
    "kyuubiki.agent-operator-task-failure/v1",
    "kyuubiki.control-plane-operator-task-failure/v1"
  ]

  def failure_receipts(payload) do
    payload
    |> collect_receipts([])
    |> Enum.reverse()
    |> Enum.uniq_by(&receipt_key/1)
  end

  def failure_actions(payload) do
    receipt_actions =
      payload
      |> failure_receipts()
      |> Enum.map(&get_in(&1, ["recovery", "required_action"]))

    (receipt_actions ++ recovery_action_values(payload))
    |> Enum.filter(&is_binary/1)
    |> Enum.reject(&(&1 == ""))
    |> Enum.uniq()
  end

  def recovery_summary(payload) do
    receipts = failure_receipts(payload)

    %{
      "next_action" => first_string_value(payload, "next_action"),
      "target_case_ids" => first_string_list_value(payload, "target_case_ids"),
      "blocked_case_ids" => first_string_list_value(payload, "blocked_case_ids"),
      "recovery_actions" => failure_actions(payload),
      "failure_receipt_count" => length(receipts),
      "failure_receipts" => receipts
    }
  end

  defp collect_receipts(%{} = payload, receipts) do
    receipts =
      Enum.reduce(@receipt_keys, receipts, fn key, acc ->
        case Map.get(payload, key) do
          %{} = receipt -> maybe_add_receipt(receipt, acc)
          _ -> acc
        end
      end)

    receipts = maybe_add_receipt(payload, receipts)

    payload
    |> Map.values()
    |> Enum.reduce(receipts, &collect_receipts/2)
  end

  defp collect_receipts(values, receipts) when is_list(values) do
    Enum.reduce(values, receipts, &collect_receipts/2)
  end

  defp collect_receipts(_value, receipts), do: receipts

  defp recovery_action_values(payload) do
    payload
    |> collect_recovery_actions([])
    |> Enum.reverse()
  end

  defp collect_recovery_actions(%{} = payload, actions) do
    actions =
      case Map.get(payload, "recovery_actions") do
        values when is_list(values) ->
          Enum.reduce(values, actions, fn action, acc -> [action | acc] end)

        _ ->
          actions
      end

    payload
    |> Map.values()
    |> Enum.reduce(actions, &collect_recovery_actions/2)
  end

  defp collect_recovery_actions(values, actions) when is_list(values) do
    Enum.reduce(values, actions, &collect_recovery_actions/2)
  end

  defp collect_recovery_actions(_value, actions), do: actions

  defp first_string_value(%{} = payload, key) do
    case Map.get(payload, key) do
      value when is_binary(value) and value != "" ->
        value

      _ ->
        payload
        |> Map.values()
        |> Enum.find_value(&first_string_value(&1, key))
    end
  end

  defp first_string_value(values, key) when is_list(values),
    do: Enum.find_value(values, &first_string_value(&1, key))

  defp first_string_value(_value, _key), do: nil

  defp first_string_list_value(%{} = payload, key) do
    case Map.get(payload, key) do
      values when is_list(values) ->
        Enum.filter(values, &is_binary/1)

      _ ->
        payload
        |> Map.values()
        |> Enum.find_value([], fn child ->
          case first_string_list_value(child, key) do
            [] -> nil
            values -> values
          end
        end)
    end
  end

  defp first_string_list_value(values, key) when is_list(values) do
    Enum.find_value(values, [], fn child ->
      case first_string_list_value(child, key) do
        [] -> nil
        values -> values
      end
    end)
  end

  defp first_string_list_value(_value, _key), do: []

  defp maybe_add_receipt(%{"schema_version" => schema} = receipt, receipts)
       when schema in @receipt_schemas do
    [receipt | receipts]
  end

  defp maybe_add_receipt(_value, receipts), do: receipts

  defp receipt_key(receipt) do
    {
      Map.get(receipt, "schema_version"),
      Map.get(receipt, "failure_stage"),
      Map.get(receipt, "reason_code"),
      Map.get(receipt, "task_id"),
      Map.get(receipt, "operator_id"),
      Map.get(receipt, "task_digest")
    }
  end
end
